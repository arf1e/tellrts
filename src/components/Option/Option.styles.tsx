import EStyleSheet from 'react-native-extended-stylesheet';

export default EStyleSheet.create({
  btn: {
    padding: 12,
    borderWidth: 1,
    borderRadius: 4,
    borderColor: '$secondary',
  },

  icon: {
    color: '$primary',
  },

  btnIcon: {
    flexDirection: 'row',
    alignItems: 'center',
  },

  btnActive: {
    borderColor: '$primary',
  },

  text: {
    color: '$gray',
    fontFamily: 'Roboto-Regular',
    fontSize: '$bcSmall',
  },

  textActive: {
    color: '$black',
  },

  btnIconText: {
    marginLeft: 16,
  },
});
