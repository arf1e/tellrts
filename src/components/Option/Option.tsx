import React, {useEffect} from 'react';
import Reanimated, {
  interpolate,
  interpolateColor,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated';
import {Pressable, PressableProps} from 'react-native';
import {BodyCopy} from '../Typography';

import styles from './Option.styles';
import colors from '../../utils/colors';
import animationConstants from '../../utils/animationConstants';

const AnimatedPressable = Reanimated.createAnimatedComponent(Pressable);

type Props = {
  title: string;
  onPress?: () => void;
  isActive?: boolean;
  pressable?: PressableProps;
};

export default ({title, onPress, isActive, pressable}: Props) => {
  const colorChange = useSharedValue(0);
  const scaleChange = useSharedValue(0);

  useEffect(() => {
    if (!isActive) {
      colorChange.value = withTiming(0, {
        duration: animationConstants.BUTTON_IN,
      });
    } else {
      colorChange.value = withTiming(1, {
        duration: animationConstants.BUTTON_OUT,
      });
    }
  }, [isActive, colorChange]);

  const animatedBtnStyle = useAnimatedStyle(
    () => ({
      borderColor: interpolateColor(
        colorChange.value,
        [0, 1],
        [colors.secondary, colors.primary],
      ),
      padding: interpolate(colorChange.value, [0, 1], [12, 11]),
      transform: [
        {
          scale: interpolate(scaleChange.value, [0, 1], [1, 0.99]),
        },
      ],
    }),
    [colorChange, scaleChange],
  );

  const onPressIn = () => {
    scaleChange.value = withTiming(1, {duration: animationConstants.BUTTON_IN});
  };

  const onPressOut = () => {
    scaleChange.value = withTiming(0, {
      duration: animationConstants.BUTTON_OUT,
    });
  };

  const btnStyle = [styles.btn, animatedBtnStyle, pressable?.style];
  const textStyle = [styles.text, isActive && styles.textActive];

  return (
    <AnimatedPressable
      {...pressable}
      style={btnStyle}
      onPress={onPress}
      onPressIn={onPressIn}
      onPressOut={onPressOut}>
      <BodyCopy style={textStyle}>{title}</BodyCopy>
    </AnimatedPressable>
  );
};
