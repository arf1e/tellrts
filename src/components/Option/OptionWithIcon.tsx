import React, {useEffect} from 'react';
import {Pressable} from 'react-native';
import Reanimated, {
  interpolate,
  interpolateColor,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated';
import {BodyCopy} from '../Typography';
import Icon from 'react-native-vector-icons/Ionicons';

import styles from './Option.styles';
import animationConstants from '../../utils/animationConstants';
import colors from '../../utils/colors';

const ReanimatedPressable = Reanimated.createAnimatedComponent(Pressable);
const ReanimatedIcon = Reanimated.createAnimatedComponent(Icon);

type Props = {
  title: string;
  icon: {
    title: string;
    size: number;
  };
  onPress: () => any;
  isActive?: boolean;
  containerStyle?: object;
};

export default ({title, icon, onPress, containerStyle, isActive}: Props) => {
  const colorChange = useSharedValue(0);
  const scaleChange = useSharedValue(0);

  useEffect(() => {
    if (!isActive) {
      colorChange.value = withTiming(0, {
        duration: animationConstants.BUTTON_IN,
      });
    } else {
      colorChange.value = withTiming(1, {
        duration: animationConstants.BUTTON_OUT,
      });
    }
  }, [isActive, colorChange]);

  const animatedBtnStyle = useAnimatedStyle(
    () => ({
      borderColor: interpolateColor(
        colorChange.value,
        [0, 1],
        [colors.secondary, colors.primary],
      ),
      borderWidth: interpolate(colorChange.value, [0, 1], [1, 2]),
      padding: interpolate(colorChange.value, [0, 1], [12, 11]),
      transform: [
        {
          scale: interpolate(scaleChange.value, [0, 1], [1, 0.99]),
        },
      ],
    }),
    [colorChange, scaleChange],
  );

  const animatedIconStyle = useAnimatedStyle(
    () => ({
      color: colors.primary,
      opacity: interpolate(colorChange.value, [0, 1], [0.7, 1]),
    }),
    [colorChange],
  );

  const onPressIn = () => {
    scaleChange.value = withTiming(1, {duration: animationConstants.BUTTON_IN});
  };

  const onPressOut = () => {
    scaleChange.value = withTiming(0, {
      duration: animationConstants.BUTTON_OUT,
    });
  };

  const btnStyle = [
    styles.btn,
    styles.btnIcon,
    containerStyle,
    animatedBtnStyle,
  ];
  const iconStyle = [styles.icon, animatedIconStyle];
  const textStyle = [
    styles.text,
    styles.btnIconText,
    isActive && styles.textActive,
  ];
  return (
    <ReanimatedPressable
      style={btnStyle}
      onPressIn={onPressIn}
      onPressOut={onPressOut}
      onPress={onPress}>
      <ReanimatedIcon style={iconStyle} name={icon.title} size={icon.size} />
      <BodyCopy style={textStyle}>{title}</BodyCopy>
    </ReanimatedPressable>
  );
};
