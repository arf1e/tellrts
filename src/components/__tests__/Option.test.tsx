import React from 'react';
import {render} from '@testing-library/react-native';

import Option from '../Option/Option';

test('Option works as expected', async () => {
  const onPressMock = jest.fn();
  const view = render(<Option title="Test Option" onPress={onPressMock} />);
  expect(view.toJSON()).toMatchSnapshot();
});
