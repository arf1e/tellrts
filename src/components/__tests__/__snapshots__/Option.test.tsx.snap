// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option works as expected 1`] = `
<View
  accessible={true}
  animatedStyle={
    Object {
      "value": Object {
        "borderColor": "rgba(188, 216, 221, 1)",
        "padding": 12,
        "transform": Array [
          Object {
            "scale": 1,
          },
        ],
      },
    }
  }
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "rgba(188, 216, 221, 1)",
      "padding": 12,
      "transform": Array [
        Object {
          "scale": 1,
        },
      ],
    }
  }
>
  <Text
    style={
      Array [
        undefined,
        Array [
          undefined,
          undefined,
        ],
      ]
    }
  >
    Test Option
  </Text>
</View>
`;
